Run whole test/Base_Tests with NI

JaroslavTulach opened the pull request:
### Pull Request Description

Fixes #10126 by various changes to make more `test/Base_Tests` pass in the Enso mode.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- [x] Test whole `test/Base_Tests` via `./runner` in CI

JaroslavTulach committed:
Try to run the whole test/Base_Tests

JaroslavTulach committed:
Catch NI error

JaroslavTulach commented:
We can run whole `./runner --run test/Base_Tests` and **not crash** now. That's marks the first milestone:
- tests is still something to work on as there is [416 failed tests](https://github.com/enso-org/enso/actions/runs/9544492501/job/26303286510?pr=10296#step:7:20149)

after 20f9589 fixes we are down to 
- [302 failed tests](https://github.com/enso-org/enso/actions/runs/9546617280/job/26309870397?pr=10296#step:7:15697)

after d61e5d2 we have
- [210 failed tests](https://github.com/enso-org/enso/actions/runs/9549033865/job/26317660718?pr=10296#step:7:11860)

finally (for today) with 501c07d we are down to
- [60 test failures](https://github.com/enso-org/enso/actions/runs/9550123136/job/26321346597?pr=10296#step:7:5176)

JaroslavTulach committed:
Instance methods on Decoder and Encoder are being called

JaroslavTulach committed:
java.util.ArrayList is often used in tests - register it for reflection

JaroslavTulach committed:
Allow access to instance methods of these classes

Akirathan commented on distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Array_Like_Helpers.enso:
```diff
@@ -29,6 +29,9 @@ from project.Data.Index_Sub_Range import drop_helper, Index_Sub_Range, take_help
 from project.Data.Ordering import Comparable
 from project.Data.Range.Extensions import all
 
+# often used in tests
```
As discussed some time ago, can we put all the `polyglot java import`s in some special file (maybe something like `test/Base_Tests/src/Polyglot_Imports.enso`?) that will contain all these imports that are necessary only to make the native-image work? The [EnsoLibraryFeature](https://github.com/enso-org/enso/blob/16c1b74218f6ca15f7a2d9dc96356d67c39e499a/engine/runner/src/main/java/org/enso/runner/EnsoLibraryFeature.java) can pick that file and register all the mentioned classes for reflection.

radeusgd replied:
Why should that be a special file? What is wrong about discovering the imports from various places in the project?

Akirathan replied:
Because the imports introduced in this PR, along with others introduced with the [previous NI-related PR](https://github.com/enso-org/enso/pull/9997) are here only to fix running of `test/Base_Tests` with the NI generated engine-runner. Running `test/Base_Tests` with the standard JVM engine-runner works even without these imports. This makes these imports special. Therefore, I would rather see them in a separate file with a comment explaining that these imports are only necessary to make the `EnsoLibraryFeature` work.

radeusgd replied:
That is a very good idea! I have misunderstood you - I thought you wanted to move all kinds of Java imports into a single file - which seemed strange.

But I understand now that you wanted to move only 'optional' imports there, right? (i.e. imports not needed for `Base` but needed for tests)

JaroslavTulach replied:
Yes, special file was the solution in https://github.com/enso-org/enso/pull/9997/commits/aaa5e9ec9ec39b510e977cfa67baa7f649b0d2c5

JaroslavTulach replied:
Fixed in dc383710b5b901808d53a8a5c0b23317b67f718e

JaroslavTulach committed:
Need reflective access to AdjustOp nested class

JaroslavTulach committed:
Need to call instance methods on LinearModel

JaroslavTulach committed:
Need access to nested classes

JaroslavTulach committed:
Put all extra files to Extra_Imports.enso

JaroslavTulach committed:
locate enso_parser library even in ./runner mode

JaroslavTulach committed:
Need to access message method

JaroslavTulach committed:
Execute ./runner --run <selected file>

JaroslavTulach committed:
./runner must embed all locales

JaroslavTulach committed:
Making sure more missing symbols is accesible

kazcw commented on lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java:
```diff
@@ -24,26 +24,28 @@ private static void initializeLibraries() {
+    } catch (NullPointerException | IllegalArgumentException | LinkageError e) {
       if (!searchFromDirToTop(e, root, "target", "rust", "debug", name)) {
-        throw new IllegalStateException("Cannot load parser from " + parser, e);
+        throw new IllegalStateException("Cannot load parser from " + whereAmI, e);
```
Should this be `root`?

JaroslavTulach replied:
That's true. Fixed in 0d902b2108. Thank you.

kazcw reviewed (approved):
Reviewed only Parser change

hubertp reviewed (approved):
It's mildly annoying that we all have to pay the price of importing additional symbols that are not being used in regular runs just because of `Base_Tests`

JaroslavTulach committed:
Replacing usage of asGuestObject in EnsoFile by own TruffleObject implementations

JaroslavTulach committed:
Support reading into a byte[]

JaroslavTulach committed:
I/O needs to be hidden behind @TruffleBoundary

JaroslavTulach committed:
Cannot be loaded from "root"

JaroslavTulach committed:
XML.from_file requires -1 at the end of the file

JaroslavTulach committed:
Include jdk.localedata to make I18N available

JaroslavTulach commented:
With cae2521faa the [conversion failure](https://github.com/enso-org/enso/actions/runs/9566749241/job/26372777530?pr=10296#step:7:7480):
```
  - [FAILED] should format correctly [2ms]
               Reason: '123,456,789.87654' did not equal '123.456.789,87654' (at /runner/_work/enso/enso/test/Base_Tests/src/Data/Decimal_Spec.enso:714:13-75).
```
goes away and we have just [82 failed tests](https://github.com/enso-org/enso/actions/runs/9567728653/job/26376172910?pr=10296#step:7:7416) - it is more than yesterday (we have problems [converting EnsoInputStream](https://github.com/enso-org/enso/actions/runs/9564840375/job/26366476053#step:7:7371) to hosted `java.io.InputStream`), but it is certainly less than [earlier today](https://github.com/enso-org/enso/actions/runs/9564840375/job/26366476053#step:7:7511).

JaroslavTulach commented:
> we have problems [converting EnsoInputStream](https://github.com/enso-org/enso/actions/runs/9564840375/job/26366476053#step:7:7371) to hosted `java.io.InputStream`), 

This is a particularly interesting problem. 

We want to create something that looks like `java.io.InputStream` in the engine builtin (that's the **engine level Java**), we want to let Enso programs access that object as it has all `InputStream` methods - that's done by [EnsoInputStream](https://github.com/enso-org/enso/pull/10296/files#diff-fddd54e102ac240a691a0dbd8e14ebff3424e311df7f106c6fc26bfa17afb57cR179) which extends [TruffleObject](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/TruffleObject.html), implements [InteropLibrary](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropLibrary.html) and exposes `invokeMember` message for `read` & co. This **works as well**.

However then we want to pass this object into **libraries Java level** (the parser API expects an `InputStream`). However that's not the same Java level as the engine! Moreover, in the NI mode those can even be **different JVMs** - one _native image_ compiled and one _hotspot JVM_ or _Espresso_! In such a setup: What looks like `InputStream` on the **engine level** is just an opaque reference on the **libraries level**.

**Update**: **Fixed** by 0695d1b5be

PS: It worked so far (as implemented by @hubertp) with `asGuestObject` because both engine and libraries were running in the same JVM - e.g. the `InputStream` reference was not opaque. But with [NI+JVM mode](https://github.com/orgs/enso-org/discussions/10121#discussioncomment-9593539) (as outlined in the **3rd party libraries** section) we cannot get the parsing working this way.

JaroslavTulach committed:
Less polyglot java imports in XML.enso

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/XML.enso:
```diff
@@ -25,19 +25,18 @@ from project.Data.Range.Extensions import all
-polyglot java import java.io.StringReader
 polyglot java import java.lang.Exception as JException
-polyglot java import javax.xml.parsers.DocumentBuilder
-polyglot java import javax.xml.parsers.DocumentBuilderFactory
```
I am [removing these `polyglot java import` statements in favor of doing more work in Java](https://github.com/enso-org/enso/pull/10296/commits/30ce7c33db372f0d42a6dce5a8911f12f6232a55), @AdRiley, @radeusgd. This is aligned with the [plan to use static methods more](https://github.com/enso-org/enso/pull/9997#discussion_r1606033800). Please keep that in mind when introducing new `polyglot java import` (especially from JDK classes) - it has its cost!

JaroslavTulach replied:
With 0695d1b5be the cost of `polyglot java import` of Java `interface` is even higher - it is automatically registered for _proxy creation_ - another reason to **expose only static methods classes** and not `org.w3c.dom.Node`, `org.w3c.dom.Element`, `Attr`, etc.

JaroslavTulach replied:
- reported as #10325

JaroslavTulach committed:
InputStreamLike interface to 'view' EnsoFile's 'java' stream as having read(byte[], int, int) method

JaroslavTulach commented:
With 0695d1b the XML `InputStream` issue is fixed and we are down to
- [56 failures](https://github.com/enso-org/enso/actions/runs/9579567799/job/26412268817?pr=10296#step:7:5820) - lowest number seen so far
- [40 failures](https://github.com/enso-org/enso/actions/runs/9583704570/job/26425595466?pr=10296#step:7:5590) after solving the `Charset` issue in d03ba23

JaroslavTulach committed:
LocalDateTime and Function are used in Base_Tests

JaroslavTulach committed:
Add all Charsets

JaroslavTulach committed:
Don't need index to runEngineDistribution

JaroslavTulach committed:
Let the Stream_Utils handle asInputStream conversion

JaroslavTulach committed:
Additional imports

JaroslavTulach committed:
Introducing File_Utils to encapsulate matcher

JaroslavTulach committed:
Avoid Python None check when Python is missing

JaroslavTulach committed:
asOutputStream wrapper

JaroslavTulach committed:
Need to support available() to make Reporting_Stream_Decoder_Spec succeed

JaroslavTulach commented on build.sbt:
```diff
@@ -3567,7 +3571,7 @@ ThisBuild / buildEngineDistributionNoIndex := {
   inputKey[Unit]("Run or --debug the engine distribution with arguments")
 runEngineDistribution := {
-  buildEngineDistribution.value
+  buildEngineDistributionNoIndex.value
```
Once this is integrated, the `runEngineDistribution` command is going to be way faster. Thanks @Akirathan for the idea.

JaroslavTulach committed:
Register OutputStreamLike for reflection

JaroslavTulach committed:
Make HttpClient.Builder ready for reflective access

JaroslavTulach committed:
Survive when js_exception (provided via asGuestValue) doesn't have the necessary attributes in NI mode

JaroslavTulach commented:
Another day and the test fight goes on:
- down to [30 failures](https://github.com/enso-org/enso/actions/runs/9592308286/job/26450626315?pr=10296#step:7:5553) with ec0918f
- down to [26 failures](https://github.com/enso-org/enso/actions/runs/9592929901/job/26452385883?pr=10296#step:7:5300) with 6ab352e, but `test/Table_Tests` seem to be broken
- just [7 failures](https://github.com/enso-org/enso/actions/runs/9595230917/job/26459502548?pr=10296#step:7:4783) with cc14d91
- last [two tests failing](https://github.com/enso-org/enso/actions/runs/9596690773/job/26464104696?pr=10296#step:7:4478) after 4c2089f

It's the final count down!

JaroslavTulach committed:
Rather call static method than Predicate.test

JaroslavTulach committed:
When XML_Document.Value is private, allocate with factory method

JaroslavTulach committed:
Include org.enso.base_test_helpers on classpath when building native executable

JaroslavTulach committed:
Another place to asOutputStream

JaroslavTulach committed:
Calling overloaded Path.of has some issues. Use File_Utils wrapper

JaroslavTulach committed:
Few more extra imports

JaroslavTulach committed:
Regenerated SubstrateVM metadata with:
JAVA_OPTS="-agentlib:native-image-agent=config-merge-dir=./engine/runner/src/main/resources/META-INF/native-image/org/enso/runner" ./built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/enso --run test/Base_Tests

JaroslavTulach committed:
Search from cwd is important for development setup

JaroslavTulach committed:
Relaxing the test as NI cannot provide location of the error (yet)

JaroslavTulach committed:
HttpClient.Version is being access

JaroslavTulach committed:
Manual tweaks to reflect-config.json file

JaroslavTulach committed:
Have to absolutize the file path

JaroslavTulach committed:
Avoid asGuestValue in EnsoProjectNode

JaroslavTulach committed:
Need to read statusCode from EnsoHttpResponse

JaroslavTulach committed:
Test is accessing IntHolder value and boxed fields

JaroslavTulach committed:
Merging with develop

JaroslavTulach committed:
Re-adding @Tail_Call as #9251 is fixed

JaroslavTulach committed:
test/Base_Tests shall be executed with assertions enabled

JaroslavTulach committed:
HTTP Response is using a lot of Java classes

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso:
```diff
@@ -45,7 +45,7 @@ type Project_Description
        - prim_root_file: The primitive root file of the project.
        - prim_config: The primitive config of the project.
-    private Value prim_root_file prim_config
+    private Value prim_root_file ns:Text n:Text
```
Something for @Akirathan to review. I replaced usage of `asGuestValue` with a constructor that holds file and two `Text` values in 71cfc0d350617390f4fcb8f57f00443177316989

JaroslavTulach committed:
polyglot java import JsonLocation to be able to extract line and column number

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Json.enso:
```diff
@@ -31,6 +30,7 @@ from project.Data.Text.Extensions import all
 from project.Metadata.Choice import Option
 from project.Metadata.Widget import Single_Choice
 
+polyglot java import com.fasterxml.jackson.core.JsonLocation
```
@hubertp, because _native image_ isn't really a friend of reflective access, I tried to eliminate as much of usage of `asGuestValue` as possible. Right now we are only using it for wrapping exception (and leaking EnsoContext):

```
enso/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java
  687,23:     var guestEx = ctx.asGuestValue(ex);

enso/engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/TopLevelScope.java
  169,22:       return context.asGuestValue(context);

enso/engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoContext.java
  492,17:   public Object asGuestValue(Object obj) {
  493,24:     return environment.asGuestValue(obj);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/MinusBuiltinDateTimeNode.java
  34,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/MinusBuiltinDurationNode.java
  34,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/MinusBuiltinTimeOfDayNode.java
  34,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/NewBuiltinDateNode.java
  27,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/NewBuiltinDateTimeNode.java
  27,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/NewBuiltinTimeOfDayNode.java
  27,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/NewBuiltinTimeZoneNode.java
  27,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/ParseBuiltinTimeZoneNode.java
  35,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/PlusBuiltinDateTimeNode.java
  34,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/PlusBuiltinDurationNode.java
  34,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/date/PlusBuiltinTimeOfDayNode.java
  34,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/immutable/NewArrayProxyBuiltinArrayLikeHelpersNode.java
  34,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/CopyBuiltinFileNode.java
  36,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/CreateDirectoryBuiltinFileNode.java
  28,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/CreationTimeBuiltinFileNode.java
  27,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/DeleteBuiltinFileNode.java
  28,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/InputStreamBuiltinFileNode.java
  35,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/LastModifiedTimeBuiltinFileNode.java
  27,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/ListImmediateChildrenArrayFileNode.java
  27,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/MoveBuiltinFileNode.java
  36,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/OutputStreamBuiltinFileNode.java
  35,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/PosixPermissionsBuiltinFileNode.java
  27,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/ReadLastBytesBuiltinFileNode.java
  27,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/io/SizeBuiltinFileNode.java
  27,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/engine/runtime/target/scala-2.13/src_managed/main/org/enso/interpreter/node/expression/builtin/CreateProcessSystemNode.java
  35,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);
  38,54:       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);

enso/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/builtins/ExecuteMethodImplGenerator.java
  145,23:               "      .asGuestValue(" + qual + "." + name + "(" + paramsApplied + "));"

enso/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/builtins/NoSpecializationClassGenerator.java
  38,9:     var asGuestValue = needsGuestValueConversion(origin);
  39,9:     if (asGuestValue
  47,40:         processingEnvironment, origin, asGuestValue, varArgExpansion);

enso/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/builtins/SafeWrapException.java
  36,63:           "    throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);");

enso/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/builtins/SpecializedMethodsGenerator.java
  293,60:             methodBody.add("  return EnsoContext.get(this).asGuestValue(result);");

enso/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/Builtin.java
  256,59:    *       throw new PanicException(EnsoContext.get(this).asGuestValue(e), this);"
```
the usage of `asGuestValue` for exceptions is still needed as 1bae3745c6544b60eef4dcca9c8bd7a2065be517 shows.

Akirathan commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date_Period.enso:
```diff
@@ -11,7 +11,10 @@ polyglot java import java.time.temporal.TemporalAdjuster
 polyglot java import java.time.temporal.TemporalAdjusters
 polyglot java import java.time.temporal.TemporalUnit
 polyglot java import org.enso.base.time.CustomTemporalUnits
+polyglot java import org.enso.base.time.Date_Utils
```
Shouldn't these imports also be kept in `Extra_Imports.enso`?

JaroslavTulach replied:
My decision making process was:
- if the failure is in `*_Spec.enso` file, put the `polyglot java import` into `Extra_Imports.enso`
- if the failure is in `Standard.Base` library file, put the import into that file.

In this case the failure was in `Date_Period.enso` file, I believe.

Akirathan commented on build/build/src/engine/context.rs:
```diff
@@ -651,7 +651,7 @@ pub async fn runner_sanity_test(
     if enso_java.is_none() {
         let test_base = Command::new(&repo_root.runner)
-            .args(["--run", repo_root.test.join("Base_Tests").as_str(), "^Text"])
+            .args(["--run", repo_root.test.join("Base_Tests").as_str()])
```
Since we can now run `enso --run test/Base_Test`, why don't we get rid of the factorial_enso code above? It seems no longer necessary.

JaroslavTulach replied:
Removed in 562e611f6b

Akirathan reviewed (approved):
This is a huge step forward. Congratulations on making all the tests pass in such a short time. Will there be a follow-up PR enabling `test/Table_Tests` as well?

GregoryTravis commented on std-bits/base/src/main/java/org/enso/base/Stream_Utils.java:
```diff
@@ -2,9 +2,84 @@
+    public int read(byte[] arr, int off, int len) throws IOException;
+
+    public default int available() {
+      return 0;
```
When would this implementation of `available` be used?

JaroslavTulach replied:
The idea was:
- it is enough to offer invocable `read` member to be `InputStreamLike`
- offering  invocable `available` member  is optional
- if the `available` member isn't provided, then `available` returns `0`

I believe it really works that way, but I haven't tested it. True.

JaroslavTulach replied:
Test written in bfa60b4 with the help of `foreign js` implementation. Thanks for encouraging me to increase test coverage.

JaroslavTulach commented:
> This is a huge step forward. Congratulations on making all the tests pass in such a short time. Will there be a follow-up PR enabling `test/Table_Tests` as well?

The next step is to implement 
- #5298

-e.g. work on support for `--jvm` option, detection of compiled & 3rd party libraries as described in the #10121 discussion. From a perspective of that task, it is important to have some _"3rd party library"_ (not compiled by NI). `Standard.Table` can play such a role.

JaroslavTulach committed:
Factorial sample program is no longer needed

JaroslavTulach committed:
Removing unused variables

JaroslavTulach committed:
Including Input_Stream_Spec in the Base_Tests

JaroslavTulach committed:
Testing InputStreamLike with and without available member
